File: src/Atlasd/Battlenet/Protocols/Game/Messages/SID_DISPLAYAD.cs
using Atlasd.Battlenet.Exceptions;
using Atlasd.Daemon;
using System.IO;

namespace Atlasd.Battlenet.Protocols.Game.Messages
{
    class SID_DISPLAYAD : Message
    {
        public SID_DISPLAYAD()
        {
            Id = (byte)MessageIds.SID_DISPLAYAD;
            Buffer = new byte[0];
        }

        public SID_DISPLAYAD(byte[] buffer)
        {
            Id = (byte)MessageIds.SID_DISPLAYAD;
            Buffer = buffer;
        }

        public override bool Invoke(MessageContext context)
        {
            Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)");

            if (context.Direction != MessageDirection.ClientToServer)
                throw new GameProtocolViolationException(context.Client, $"{MessageName(Id)} must be sent from client to server");

            if (Buffer.Length < 14)
                throw new GameProtocolViolationException(context.Client, $"{MessageName(Id)} buffer must be at least 14 bytes");

            using var m = new MemoryStream(Buffer);
            using var r = new BinaryReader(m);

            var platformId = r.ReadUInt32();
            var productId = r.ReadUInt32();
            var adId = r.ReadUInt32();
            var adFilename = r.ReadByteString();
            var adUrl = r.ReadByteString();

            if (Battlenet.Common.ActiveAds.TryGetValue(adId, out var ad) && ad != null)
            {
                ad.IncrementDisplayCount();

                Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"Client displayed advertisement (id: {adId}) (filename: {ad.Filename})");
            }
            else
            {
                Logging.WriteLine(Logging.LogLevel.Warning, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"Client displayed advertisement (id: 0x{adId:X8}) but id cannot be found");
            }

            return true;
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/Messages/SID_DISPLAYAD.vb
Imports AtlasV.Battlenet.Exceptions
Imports AtlasV.Daemon
Imports System.IO

Namespace AtlasV.Battlenet.Protocols.Game.Messages
    Class SID_DISPLAYAD
        Inherits Message

        Public Sub New()
            Id = CByte(MessageIds.SID_DISPLAYAD)
            Buffer = Array.Empty(Of Byte)()
        End Sub

        Public Sub New(ByVal varBuffer As Byte())
            Id = CByte(MessageIds.SID_DISPLAYAD)
            Buffer = varBuffer
        End Sub

        Public Overrides Function Invoke(ByVal context As MessageContext) As Boolean
            Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")
            If context.Direction <> MessageDirection.ClientToServer Then Throw New GameProtocolViolationException(context.Client, $"{MessageName(Id)} must be sent from client to server")
            If Buffer.Length < 14 Then Throw New GameProtocolViolationException(context.Client, $"{MessageName(Id)} buffer must be at least 14 bytes")

            Dim platformId, productId, adId As UInt32
            Dim adFilename, adUrl As String
            Using m = New MemoryStream(Buffer)
                Using r = New BinaryReader(m)
                    platformId = r.ReadUInt32()
                    productId = r.ReadUInt32()
                    adId = r.ReadUInt32()
                    adFilename = r.ReadString()
                    adUrl = r.ReadString()
                End Using
            End Using

            Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"Ad [Id: 0x{adId}] was displayed!")
            Return True
        End Function
    End Class
End Namespace
